horovod/common/ops/nccl_operations.cc

```diff
@@ -582,6 +582,62 @@ Status NCCLBroadcast::Execute(std::vector<TensorTableEntry>& entries,
       entries, true, nccl_op_context_.error_check_callback_);
 }
 
+Status NCCLAllgather::AllocateOutput(std::vector<TensorTableEntry>& entries,
+                                   const Response& response,
+                                   int64_t**& entry_component_sizes,
```

Comment:
Why are you passing the 2d array and 1d array pointers by reference? You are not modifying the pointer address and passing them by copy has no perf penalty since sizeof(copied_arg)==size(ptrType).

Reply:
This is the existing interface for this function: https://github.com/horovod/horovod/blob/133ef0725253db83cfb82a4ed4003df76d189829/horovod/common/ops/collective_operations.h#L143-L146

For this PR, I am just adding specializations to use the events for the scheduling.